src/python/grpcio/grpc/_cython/_cygrpc/fork_posix.pyx.pxi

```diff
@@ -0,0 +1,161 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+_AWAIT_THREADS_TIMEOUT_SECONDS = 5
+
+_TRUE_VALUES = ['yes',  'Yes',  'YES', 'true', 'True', 'TRUE', '1']
+
+# This flag enables experimental support within gRPC Python for applications
+# that will fork() without exec(). When enabled, gRPC Python will attempt to
+# pause all of its internally created threads before the fork syscall proceeds.
+#
+# For this to be successful, the application must not have multiple threads of
+# its own calling into gRPC when fork is invoked. Any callbacks from gRPC
+# Python-spawned threads into user code (e.g., callbacks for asynchronous RPCs)
+# must  not block and should execute quickly.
+#
+# This flag is not supported on Windows.
+_GRPC_ENABLE_FORK_SUPPORT = (
+    os.environ.get('GRPC_ENABLE_FORK_SUPPORT', '0')
+        .lower() in _TRUE_VALUES)
+
+_GRPC_POLL_STRATEGY = os.environ.get('GRPC_POLL_STRATEGY')
+
+cdef void __prefork() nogil:
+    with gil:
+        with _fork_state.fork_in_progress_condition:
+            _fork_state.fork_in_progress = True
+        if not _fork_state.active_thread_count.await_zero_threads(
+                _AWAIT_THREADS_TIMEOUT_SECONDS):
+            _LOGGER.error(
+                'Failed to shutdown gRPC Python threads prior to fork. '
+                'Behavior after fork will be undefined.')
+
+
+cdef void __postfork_parent() nogil:
+    with gil:
+        with _fork_state.fork_in_progress_condition:
+            _fork_state.post_fork_child_cleanup_callbacks = []
+            _fork_state.fork_in_progress = False
+            _fork_state.fork_in_progress_condition.notify_all()
+
+
+cdef void __postfork_child() nogil:
+    with gil:
+        with _fork_state.fork_in_progress_condition:
+            for state_to_reset in _fork_state.postfork_states_to_reset:
+                state_to_reset.reset_postfork_child()
+            _fork_state.fork_epoch += 1
+            _fork_state.post_fork_child_cleanup_callbacks = []
+            _fork_state.fork_in_progress = False
+
+
+def fork_handlers_and_grpc_init():
+    grpc_init()
+    if _GRPC_ENABLE_FORK_SUPPORT:
+        # TODO(ericgribkoff) epoll1 is default for grpcio distribution. Decide whether to expose
+        # grpc_get_poll_strategy_name() from ev_posix.cc to get actual polling choice.
+        if _GRPC_POLL_STRATEGY is not None and _GRPC_POLL_STRATEGY != "epoll1":
+            _LOGGER.error(
+                'gRPC Python fork support is only compatible with the epoll1 '
+                'polling engine')
+            return
+        with _fork_state.fork_handler_registered_lock:
+            if not _fork_state.fork_handler_registered:
+                pthread_atfork(&__prefork, &__postfork_parent, &__postfork_child)
+                _fork_state.fork_handler_registered = True
+
+
+def fork_managed_thread(target, args=()):
+    if _GRPC_ENABLE_FORK_SUPPORT:
+        def managed_target(*args):
+            _fork_state.active_thread_count.increment()
+            target(*args)
```

Comment:
Yes, good catch. Fixed

Reply:
There's another bug here: the active thread count is only incremented once the thread's target function has actually started to run. This is problematic in (at least) the following concrete scenario: a call to channel.subscribe() starts a new thread and sets channel._connectivity_state.polling=True. Fork is invoked before the new thread's first line of code is actually run, and so the active thread count hasn't been incremented. This means the thread's code could go on to do bad things itself (such as grab a lock) before fork actually executes, and also that the child will inherit the channel with the belief that it already has a connectivity polling thread, when it does not.

I addressed this by introducing cygrpc.ForkManagedThread as a new delegating class (not an override of threading.Thread), whose start() method increments the active thread count before calling threading.Thread.start(). This will increment the global thread count before the call to start() returns, meaning that any subsequent calls to fork will correctly wait for this thread to pause/conclude before proceeding.